internal/icinga2/client.go

```diff
@@ -0,0 +1,388 @@
+// be cached until the catch-up-phase has finished, while replayed events will be delivered directly.
+//
+// Communication takes place over the eventDispatcherEventStream and catchupPhaseRequest channels.
+func (client *Client) worker() {
+	var (
+		// catchupEventCh either emits events generated during the catch-up-phase from catch-up-workers or one final
+		// error if something went wrong. It will be closed when catching up is done, which indicates the select below
+		// to switch phases. When this variable is nil, this Client is in the normal operating phase.
+		catchupEventCh chan *catchupEventMsg
+		// catchupCancel cancels, if not nil, all running catch-up-workers, e.g., when restarting catching-up.
+		catchupCancel context.CancelFunc
+
+		// catchupBuffer holds Event Stream events to be replayed after the catch-up-phase has finished.
+		catchupBuffer = make([]*event.Event, 0)
+		// catchupCache maps event.Events.Name to API time to skip replaying outdated events.
+		catchupCache = make(map[string]time.Time)
+
+		// catchupFailCounter indicates how many prior catch-up-phase attempts have failed. It will be used to
+		// rate limit catch-up-phase restarts.
+		catchupFailCounter int
+	)
+
+	// catchupReset resets all catchup variables to their initial empty state.
+	catchupReset := func() {
+		catchupEventCh, catchupCancel = nil, nil
+		catchupBuffer = make([]*event.Event, 0)
+		catchupCache = make(map[string]time.Time)
+	}
+
+	// catchupCacheUpdate updates the catchupCache if this eventMsg seems to be the latest of its kind.
+	catchupCacheUpdate := func(ev *eventMsg) {
+		ts, ok := catchupCache[ev.event.Name]
+		if !ok || ev.apiTime.After(ts) {
+			catchupCache[ev.event.Name] = ev.apiTime
+		}
+	}
+
+	for {
+		select {
+		case <-client.Ctx.Done():
+			client.Logger.Warnw("Closing down main worker as context is finished", zap.Error(client.Ctx.Err()))
+			return
+
+		case <-client.catchupPhaseRequest:
+			if catchupEventCh != nil {
+				client.Logger.Warn("Switching to catch-up-phase was requested while already catching up, restarting phase")
+				catchupCancel()
+			}
+
+			client.Logger.Info("Worker enters catch-up-phase, start caching up on Event Stream events")
+			catchupReset()
+			catchupEventCh, catchupCancel = client.startCatchupWorkers(
+				min(3*time.Minute, time.Duration(math.Exp2(float64(catchupFailCounter))-1)*time.Second))
+
+		case catchupMsg, ok := <-catchupEventCh:
+			// Process an incoming event
+			if ok && catchupMsg.error == nil {
+				client.CallbackFn(catchupMsg.eventMsg.event)
+				catchupCacheUpdate(catchupMsg.eventMsg)
+				break
+			}
+
+			// Abort and restart the catch-up-phase when receiving an error.
+			if ok && catchupMsg.error != nil {
+				client.Logger.Warnw("Worker leaves catch-up-phase with an error, another attempt will be made", zap.Error(catchupMsg.error))
+				go func() {
+					select {
+					case <-client.Ctx.Done():
+					case client.catchupPhaseRequest <- struct{}{}:
+					}
+				}()
```

Comment:
IMO a bit strange to start a go routine that only has the purpose to send an event to the very event loop that started it. Also, it intentional that `catchupReset()` is done twice?

An alternative would be to just do the `client.startCatchupWorkers()` here as well. Also shouldn't this here be the only place then where an delay is needed? In that case, you wouldn't even have to duplicate that delay.

Maybe it's also an option to signal a restart due to an error via `client.catchupPhaseRequest` directly instead of `catchupEventCh`. (To be fair, I haven't fully thought this through, so that might result in other strange things.)